crates/wasmtime/src/runtime/store.rs

```diff
@@ -1334,6 +1369,93 @@ impl StoreOpaque {
         &self.vm_store_context
     }
 
+    #[inline]
+    pub fn record_buffer_mut(&mut self) -> Option<&mut RecordBuffer> {
+        self.record_buffer.as_mut()
+    }
+
+    #[inline]
+    pub fn replay_buffer_mut(&mut self) -> Option<&mut ReplayBuffer> {
+        self.replay_buffer.as_mut()
+    }
+
+    /// Record the given event into the store's record buffer
+    ///
+    /// Convenience wrapper around [`Recorder::record_event`]
+    #[inline]
+    pub(crate) fn record_event<T, F>(&mut self, f: F)
+    where
+        T: Into<RREvent>,
+        F: FnOnce(&RecordMetadata) -> T,
+    {
+        if let Some(buf) = self.record_buffer_mut() {
+            buf.record_event(f);
+        }
+    }
+
+    /// Conditionally record the given event into the store's record buffer
+    ///
+    /// Convenience wrapper around [`Recorder::record_event_if`]
+    #[inline]
+    pub(crate) fn record_event_if<T, P, F>(&mut self, pred: P, f: F)
```

Comment:
It seems that the `_if` part of this combinator is mainly used to conditionalize running the closure on `add_validation`. Would it make sense to instead have `record_event` and `record_event_validation` entry points, or perhaps even just wrap the relevant callsites in `if r.add_validation { store.record_event(...); }`? In general I'm wary of complex type-generic closure-taking helpers like this unless they pay their way to a significant degree.